Added JSON output format for baseline cmdlets and more refactoring

ArmaanMcleod opened the pull request:
## PR Summary

Fix #839

Also made some slight improvements:

- Refactored Yaml implementation to sort properties consistently across the output. Some types such as dictionaries and arrays were not sorting due to not sorting the inner types. Made the same changes to JSON.
- Added missing `Output.JsonIndent` option to README
- Split the Yaml/Json mapping classes, although was trying to find a way to consolidate these implementations into one. Probably is possible but might get messy.
- Converted alot of variables to `var` to stay consistent with codebase. 

### Sample Output

```powershell
Get-PSRuleBaseline -OutputFormat Json -Module PSRule.Rules.Azure -Option @{'Output.JsonIndent' = 4}
```

```jsonc
[
    {
        // Synopsis: Default baseline for Azure rules.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "name": "Azure.Default"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA"
                }
            }
        }
    },
    {
        // Synopsis: Includes Azure features in preview.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "name": "Azure.Preview"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": [
                        "GA",
                        "preview"
                    ]
                }
            }
        }
    },
    {
        // Synopsis: Includes all Azure rules.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "name": "Azure.All"
        },
        "spec": {
            "rule": {
                "tag": {}
            }
        }
    },
    {
        // Synopsis: Include rules released June 2020 or prior for Azure GA features.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "annotations": {
                "obsolete": "true"
            },
            "name": "Azure.GA_2020_06"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA",
                    "ruleSet": [
                        "2020_06"
                    ]
                }
            }
        }
    },
    {
        // Synopsis: Include rules released September 2020 or prior for Azure GA features.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "annotations": {
                "obsolete": "true"
            },
            "name": "Azure.GA_2020_09"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA",
                    "ruleSet": [
                        "2020_06",
                        "2020_09"
                    ]
                }
            }
        }
    },
    {
        // Synopsis: Include rules released December 2020 or prior for Azure GA features.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "annotations": {
                "obsolete": "true"
            },
            "name": "Azure.GA_2020_12"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA",
                    "ruleSet": [
                        "2020_06",
                        "2020_09",
                        "2020_12"
                    ]
                }
            }
        }
    },
    {
        // Synopsis: Include rules released March 2021 or prior for Azure GA features.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "annotations": {
                "obsolete": "true"
            },
            "name": "Azure.GA_2021_03"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA",
                    "ruleSet": [
                        "2020_06",
                        "2020_09",
                        "2020_12",
                        "2021_03"
                    ]
                }
            }
        }
    },
    {
        // Synopsis: Include rules released June 2021 or prior for Azure GA features.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "annotations": {
                "obsolete": "true"
            },
            "name": "Azure.GA_2021_06"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA",
                    "ruleSet": [
                        "2020_06",
                        "2020_09",
                        "2020_12",
                        "2021_03",
                        "2021_06"
                    ]
                }
            }
        }
    },
    {
        // Synopsis: Include rules released September 2021 or prior for Azure GA features.
        "apiVersion": "github.com/microsoft/PSRule/v1",
        "kind": "Baseline",
        "metadata": {
            "name": "Azure.GA_2021_09"
        },
        "spec": {
            "rule": {
                "tag": {
                    "release": "GA",
                    "ruleSet": [
                        "2020_06",
                        "2020_09",
                        "2020_12",
                        "2021_03",
                        "2021_06",
                        "2021_09"
                    ]
                }
            }
        }
    }
]
```

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Code changes**
  - [x]Have unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Microsoft/PSRule/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

ArmaanMcleod commented on src/PSRule/Pipeline/Output/JsonOutputWriter.cs:
```diff
@@ -22,16 +27,50 @@ protected override string Serialize(object[] o)
+                // Workaround given JSON.NET doesn't support writing inline comments
+                // Can write multiline comments first then replace with inline comment delimiter
+                // We use /\*(.*?)\*/ to capture comments between /* */ deimiters
+                var jsonWithMultilineCommentsReplaced = Regex.Replace(jsonString, @"/\*(.*?)\*/", match =>
```
This was the workaround I did to write `//` comments. The `JsonWriter.WriteComment` method in JSON.Net only writes multiline `/* */` comments. Let me know if you find a better way. This seems like a hack, but works I guess 😄.

BernieWhite reviewed (changes requested):
@ArmaanMcleod I've added some changes instead of the regex. Have a look if you are ok with the changes.

Also we might need to update the tests. When we have a line comment `//` the following object will need to be on a separate line to avoid being commented out.

ArmaanMcleod commented:
> @ArmaanMcleod I've added some changes instead of the regex. Have a look if you are ok with the changes.
> 
> Also we might need to update the tests. When we have a line comment `//` the following object will need to be on a separate line to avoid being commented out.

@BernieWhite Thanks for the changes, definetely a much better approach to regex 😄. I wasn't very familiar with the JSON.Net API methods to override `WriteComment`. Great to see how that can be done. 

I've updated the tests as well to hand newlines for 0 indentation. Will see if this fixes the CI errors, was passing on my local machine.
